Manually add specific track to a sub-playlist

featherbear opened the pull request:
<!-- This is a template - do add to or remove from as needed -->
<!-- Please provide a general summary of your changes in the Title above 🚀 -->
## Context

[Corresponding Pivotal Tracker story](https://www.pivotaltracker.com/story/show/175033788)

## Description 💬

Allow tracks to be dragged from the master playlist, this allows new items to be added to sub-playlists.

----

_Note: A suggestion I thought of, was to create a copy of the track entry during drag (so that it doesn't affect the contents of the master playlist - however this feature is not possible with the current drag-n-drop library_

## Dependent PRs

- [x] https://github.com/Splitify/splitify/pull/33
- [x] https://github.com/Splitify/splitify/pull/34

featherbear committed:
Merge branch 'track_and_group_wrapper' into add_from_master_playlist

featherbear committed:
Merge remote-tracking branch 'origin/master' into add_from_master_playlist

featherbear committed:
Merge branch 'master' into add_from_master_playlist

featherbear committed:
Merge branch 'source_origin' into add_from_master_playlist

featherbear committed:
Merge branch 'perf/preview' into add_from_master_playlist

featherbear committed:
Enable master playlist drag

featherbear committed:
Create a copy of the track when dragged from the master playlist

featherbear committed:
Formatting

featherbear committed:
Add onFilterUpdate to MasterPlaylist and PlaylistWrapper

featherbear committed:
Wrap filter function into a callback to reduce re-render count

featherbear committed:
Change filterList store into a POJO instead of a React state

featherbear committed:
Rewrite master playlist drag

CarlinWilliamson reviewed (changes requested):
If you have two subplaylists (A and B) with the same genres, drag a song from the master playlist into A. You can no longer drag between A and B

Single subplaylist which includes song A. If you move the corresponding song B (same song as B) from the master playlist into that subplaylist _below_ song A then try to move it up, both A and B move

featherbear committed:
Fix inconsistent consistent state

featherbear committed:
Merge remote-tracking branch 'origin/master' into add_from_master_playlist

featherbear committed:
Focus on specific track if custom

Will break for a playlist that contains two entries of the same track that aren't right next to each other

featherbear committed:
Refactor Track creation

* Fix `expand` helper
* Initialise all Tracks as PlaylistTrack
* Remove `unsafe` parameter from `asPlaylistTrack`
* Add `clone` function

featherbear committed:
Clone a track on drag from master playlist

featherbear committed:
Use UUID as the Draggable ID

featherbear committed:
Cleanup

featherbear committed:
Fix UI refresh

featherbear commented:
> 
> 
> If you have two subplaylists (A and B) with the same genres, drag a song from the master playlist into A. You can no longer drag between A and B

Fixed in 8e7a9bd4c03a82d837b113193537032119da294b

> 
> Single subplaylist which includes song A. If you move the corresponding song B (same song as B) from the master playlist into that subplaylist _below_ song A then try to move it up, both A and B move

Fixed in 8bcdd009ee3a8c75eca6a4aa5d0ab943012b2ab7

featherbear committed:
Maintain custom tracks and sort order when changing genre and feature filter

featherbear committed:
Merge remote-tracking branch 'origin/master' into add_from_master_playlist

featherbear committed:
Merge branch 'master' into add_from_master_playlist

featherbear committed:
Remove track from source pool on original drag

featherbear committed:
Fix dragging a new track to the bottom of a sub playlist

CarlinWilliamson reviewed (approved):
lgtm. Finally :)
